src/storage/mvcc/txn.rs

```diff
@@ -381,18 +382,48 @@ impl<S: Snapshot> MvccTxn<S> {
                 last_lock_ttl = lock.ttl;
             }
         } else if is_pessimistic_lock {
-            // Pessimistic lock does not exist, the transaction should be aborted.
-            warn!(
-                "prewrite failed (pessimistic lock not found)";
-                "start_ts" => self.start_ts,
-                "key" => %key
-            );
-
-            return Err(ErrorInner::PessimisticLockNotFound {
-                start_ts: self.start_ts,
-                key: key.into_raw()?,
+            if !pipelined_pessimistic_lock {
+                // Pessimistic lock does not exist, the transaction should be aborted.
+                warn!(
+                    "prewrite failed (pessimistic lock not found)";
+                    "start_ts" => self.start_ts,
+                    "key" => %key
+                );
+                return Err(ErrorInner::PessimisticLockNotFound {
+                    start_ts: self.start_ts,
+                    key: key.into_raw()?,
+                }
+                .into());
+            }
+            if let Some((commit_ts, _)) = self.reader.seek_write(&key, TimeStamp::max())? {
+                if commit_ts < self.start_ts {
+                    // Used pipelined pessimistic lock acquiring in this txn but failed
+                    // Luckily no other txn modified this lock, amend it by overwriting.
+                    MVCC_CONFLICT_COUNTER
+                        .pipelined_acquire_pessimistic_lock_amend_update
+                        .inc();
+                } else {
+                    warn!(
+                        "prewrite failed (pessimistic lock not found)";
+                        "start_ts" => self.start_ts,
+                        "commit_ts" => commit_ts,
+                        "key" => %key
+                    );
+                    MVCC_CONFLICT_COUNTER
+                        .pipelined_acquire_pessimistic_lock_false_success
+                        .inc();
+                    return Err(ErrorInner::PessimisticLockNotFound {
+                        start_ts: self.start_ts,
+                        key: key.into_raw()?,
+                    }
+                    .into());
+                }
+            } else {
+                // Key not exists, amend it
+                MVCC_CONFLICT_COUNTER
```

Comment:
I think separate the two metric could be helpful by tell us which case is happening:
1. Insert a row in pipelined-pessimistic-txn, but meet false-success
2. Update a row in pipelined-pessimistic-txn, but meet false-success
There are a big change case-1 happens, because a lot of inserting could cause region-splitting. and case-2 should be rare.

Reply:
Actually, we can't infer whther it's an update op or insert op from the existence of `Write`.